{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pangenome Construction with Roary\n",
    "\n",
    "No exercises in this section.\n",
    "\n",
    "# The Pangenome explained\n",
    "\n",
    "## Check your understanding\n",
    "\n",
    "**Q1: The pangenome contains:**  \n",
    "a) Only genes present in one isolate in a population  \n",
    "b) All genes from all isolates in a population  \n",
    "c) Only genes present in all isolates in a population  \n",
    "  \n",
    "**A1: b) All genes from all isolates in a population**  \n",
    "\n",
    "**Q2: Core genes are:**  \n",
    "a) Often important for basic cell functions  \n",
    "b) Present in only a subset of the isolates of a population  \n",
    "c) Often related to drug resistance  \n",
    " \n",
    "**A2: a) Often important for basic cell functions**    \n",
    "\n",
    "# Preparing the input data for Roary\n",
    "\n",
    "##\n",
    "## Check your understanding\n",
    "**Q3: Why do we need to run Prokka?**  \n",
    "a) It will perform QC on our data  \n",
    "b) It will annotate our data  \n",
    "c) We don't, Roary can handle fasta files as input  \n",
    "\n",
    "**A3:** b) It will annotate our data  \n",
    "\n",
    "**Q4: Why do we use the --locustag option when we run Prokka?**  \n",
    "a) To make it easier to keep track of the output files  \n",
    "b) Because Roary won't work without it  \n",
    "c) To make the Roary results easier to interpret  \n",
    "  \n",
    "**A4:** c) To make the Roary results easier to interpret   \n",
    "\n",
    "# Performing QC on your data\n",
    "\n",
    "##\n",
    "##\n",
    "##\n",
    "##\n",
    "## Check your understanding\n",
    "\n",
    "**Q5: Why is it important to QC your data?**   \n",
    "**A5:** If the data is bad going in, the results will be bad coming out.  \n",
    "  \n",
    "**Q6: You're not getting any core genes when you run Roary. What could be the reason?**  \n",
    "**A6:** Most commonly there is contamination, or the geomes are too fragmented.  \n",
    "\n",
    "**Q7: What is the size of the assembly for sample1?**  \n",
    "**A7:** 2096319\n",
    "\n",
    "**Q8: How many contigs are in the assembly of sample1?**  \n",
    "**A8:** 38\n",
    "\n",
    "# Constructing a Pangenome with Roary\n",
    "\n",
    "##\n",
    "##\n",
    "## Check your understanding\n",
    "\n",
    "**Q9: Why do we want to run Roary with MAFFT?**  \n",
    "a) Because it's quicker than to run Roary without the -e option  \n",
    "b) To get more accurate results  \n",
    "c) To generate a core gene alignment  \n",
    "\n",
    "**A9:** c) To generate a core gene alignment\n",
    "\n",
    "**Q10: Why do we use the -p otion?**  \n",
    "a) We have to when we use MAFFT  \n",
    "b) To speed up the run  \n",
    "c) To get a nice tree\n",
    "  \n",
    "**A10:** b) To speed up the run      \n",
    "\n",
    "# Exploring the results\n",
    "\n",
    "##\n",
    "##\n",
    "##\n",
    "## Check your understanding\n",
    "\n",
    "**Q11: Approximately how many genes would you expect to see in the summary_statistics.txt file if you are working with a species with a genome size of 5,000,000 bases?**  \n",
    "a) 500  \n",
    "b) 5000  \n",
    "c) 50,000  \n",
    "\n",
    "**A11:** b) 5000    \n",
    "\n",
    "**Q12: What does the accessory_binary_genes.fa.newick file provide?**  \n",
    "a) A pylogenetic tree ready for publishing  \n",
    "b) Nothing, it is useless  \n",
    "c) A quick insight to the data  \n",
    "\n",
    "**A12:** c) A quick insight to the data    \n",
    "\n",
    "**Q13: For query_pan_genome, what option should you use to get the accessory genome?**  \n",
    "a) union    \n",
    "b) intersection  \n",
    "c) complement  \n",
    "\n",
    "**A13:** c) complement  \n",
    "\n",
    "# Visualising the results with phandango\n",
    "\n",
    "## Check your understanding\n",
    "\n",
    "**Q14: What is the name of this gene cluster?**  \n",
    "**A14:** lytN  \n",
    "\n",
    "**Q15: Is this a core gene?**  \n",
    "**A15:** No (it is only present in sample 1 and 2)\n",
    "\n",
    "# Creating genome assemblies\n",
    "\n",
    "##\n",
    "##\n",
    "## Check your understanding\n",
    "\n",
    "Note these metrics may differ slightly.\n",
    "\n",
    "**Q16: What is the size of the assembly?**  \n",
    "**A16:** 2100403\n",
    "\n",
    "**Q17: How many contigs did it assemble into?**   \n",
    "**A17:** 164\n",
    "\n",
    "**Q18: What is the largest contig?**   \n",
    "**A18:** 205299\n",
    "\n",
    "**Q19: What is the N50?**  \n",
    "**A19:** 73737\n",
    "\n",
    "**Q20: Is this a good assembly?**  \n",
    "**A20:** Yes, a reasonable quality assembly."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
